### PR TITLE
Use antsibull sphinx extension

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
The Pygments lexer and part of the CSS is contained in the antsibull sphinx extension (https://github.com/ansible-community/antsibull/tree/main/sphinx_antsibull_ext), so there's no need to duplicate these parts in the docsite sources.

##### ISSUE TYPE
- Docs Pull Request

##### COMPONENT NAME
docsite

### CODE REVIEW

**Comment by felixfontein** [comment]
This is strange. Sphinx 2.1.2 (https://github.com/ansible/ansible/blob/devel/docs/docsite/requirements.txt#L5) already has add_css_file (https://github.com/sphinx-doc/sphinx/blob/v2.1.2/sphinx/application.py#L960). Maybe an even older version of Sphinx is used here? It must have been before version 1.8.0, since add_css_file was added then (https://github.com/sphinx-doc/sphinx/commit/3afc72fba41f945610263be79d14c95d21478248).

This should get fixed by ansible-community/antsibull#233, but maybe the very old Sphinx version used for docsite testing also needs to be updated :)

**Comment by abadger** [comment]
CC @samccann about the older Sphinx requirement

**Comment by webknjaz** [comment]
@abadger FYI there's a PR allowing better sphinx: https://github.com/ansible/ansible/pull/73176.

**Comment by felixfontein** [comment]
In any case, whatever comes first (#73176 merged or new antsibull release), this issue should be fixed soon :)

**Comment by felixfontein** [comment]
Restarting tests (since rstcheck / sphinx were updated).

**Comment by felixfontein** [comment]
@samccann for some reason, the sanity test is still using Sphinx v1.7.9.

**Comment by felixfontein** [comment]
@mattclay about the requirement constraings change: is this change OK? I could also change it to `>=1.8.0,<=2.1.2`. (Alternatively this can also be fixed by creating a new version of the default test docker container which has a newer version of Sphinx, and by not modifying the constraints.)

**Comment by felixfontein** [comment]
(Also it would be great if we could backport this, so we can start making some modifications to the doc site build in antsibull that would require support by its sphinx extension.)

**Comment by mattclay** [comment]
@felixfontein I just updated the default test container. Can you confirm it has the version you're looking for?

**Comment by mattclay** [comment]
@felixfontein Major changes to requirements/constraints shouldn't be backported, to avoid introducing breaking changes.

**Comment by felixfontein** [comment]
@mattclay
```
root@89747a37ef03:~/ansible$ pip show sphinx
Name: Sphinx
Version: 2.1.2
```
looks good to me :)

About not backporting changes to the constraints: hmm that's not great, but I think we can get it work. I already added some compatibility code to the antsibull sphinx theme so it *should* work with Sphinx < 1.8, but it's not yet in a release.

**Comment by felixfontein** [comment]
@mattclay @samccann @abadger with the new test container, it now works without any further adjustments! :tada:

**Comment by felixfontein** [comment]
recheck

**Comment by samccann** [comment]
@felixfontein - I'm having trouble getting this staged on test (my fault... jenkins...grrr). Meanwhile I  did a 'make webdocs ANSIBLE_VERSION=3' locally and the parameter tables for modules are missing the table lines - 

image

I poked around at a few other non-plugin pages and it looks fine otherwise.

**Comment by felixfontein** [comment]
@samccann thanks a lot for testing! That's definitely not intended - I'll try to run the build locally myself to see what's happening (I hope I manage over the weekend :) ).

**Comment by felixfontein** [comment]
@samccann the problem was that the configuration files explicitly overwrite the CSS file list, thus disallowing extensions to add their own CSS files. I explicitly added the antsibull CSS file to the list in d4c8659c70829672b74e155a4ac18303fe26d4e7.

**Review by webknjaz — DISMISSED** [review]
This is no longer necessary — the theme has Pygments entrypoints declared so its already available to Sphinx.

**Comment by felixfontein** [comment]
> This is no longer necessary — the theme has Pygments entrypoints declared so its already available to Sphinx.

This isn't about the Pygments lexer, but about the CSS. So it's still necessary.

**Comment by webknjaz** [comment]
I believe those CSS styles are also embedded in the theme currently. So it's probably best to work towards removing the lexer from antsibull, then remove those from the theme, and only then integrate that ext here.

If we accept it as is, you'll be probably overriding the lexer that's already coming from the theme and I'm not sure how the styles would play together (needs checking). So let's try to do this gradually. WDYT?

**Comment by felixfontein** [comment]
Yes, it should only be merged once that happened. I only rebased to resolve the conflicts.

BTW, is it intentional that the ansible-core CSS (docs/docsite/_static/core.css) is still there? It contains a lot of definitions that should be part of the theme.

**Comment by webknjaz** [comment]
> BTW, is it intentional that the ansible-core CSS (docs/docsite/_static/core.css) is still there? It contains a lot of definitions that should be part of the theme.

Good catch! I must've missed that. It's not used and the color overrides are now in https://github.com/ansible/ansible/blob/devel/docs/docsite/_static/css/core-color-scheme.css.

**Comment by webknjaz** [comment]
> Yes, it should only be merged once that happened. I only rebased to resolve the conflicts.

Good. I believe there's no blockers for the antsibull cleanup atm.

**Comment by webknjaz** [comment]
Why? Is anything else using the extension? If not, it should be independent.

**Comment by felixfontein** [comment]
Antsibull itself needs it. Otherwise its docsite extra docs linter will fail for example for community.general.

**Comment by webknjaz** [comment]
> Antsibull itself needs it. Otherwise its docsite extra docs linter will fail for example for community.general.

Oh, makes sense for as long as those docs don't use the theme (yet?).

**Comment by felixfontein** [comment]
I don't think it's related to the theme? When linting `.rst` files with `rstcheck`, lexers mentioned in the `.rst` files need to be found, otherwise `rstcheck` will report errors. Not having the `ansible-output` lexer present recently caused community.general's CI to fail until it was included in antsibull.

**Comment by webknjaz** [comment]
Yes. My point is that if the project uses the theme, they'd have the lexer in their env automatically. But if it's necessary in places where the theme isn't currently used, the lexer needs to come from another place, you're right.

File: docs/docsite/sphinx_conf/2.10_conf.py [hunk 1]
```diff
@@ -43,7 +43,7 @@
 # They can be extensions
 # coming with Sphinx (named 'sphinx.ext.*') or your custom ones.
 # TEST: 'sphinxcontrib.fulltoc'
-extensions = ['sphinx.ext.autodoc', 'sphinx.ext.intersphinx', 'notfound.extension']
+extensions = ['sphinx.ext.autodoc', 'sphinx.ext.intersphinx', 'notfound.extension', 'sphinx_antsibull_ext']
```
Comment (webknjaz) [inline docs/docsite/sphinx_conf/2.10_conf.py]: How about making this multiline and adding a comment on why the last ext is needed?

File: docs/docsite/_static/core.css [hunk 2]
```diff
@@ -22,9 +22,9 @@
  * / is given the class .value-separator
  * required is given the class .value-required
  * added in x.y is given the class .value-added-in
- *//*! The extra .rst-content is so this will override rtd theme */.rst-content table.documentation-table td{vertical-align:top}table.documentation-table td:first-child{white-space:nowrap;vertical-align:top}table.documentation-table td:first-child p:first-child{font-weight:700;display:inline}/*! This is now redundant with above position-based styling *//*!
+ *//*! The extra .rst-content is so this will override rtd theme */.rst-content table.documentation-table td{vertical-align:top}table.documentation-table td:first-child p:first-child{font-weight:700;display:inline}/*! This is now redundant with above position-based styling *//*!
```
Comment (webknjaz) [inline docs/docsite/_static/core.css]: Instead of changing this file, it looks like it should be just removed.

Comment (webknjaz) [reply docs/docsite/_static/core.css]: Ref: https://github.com/ansible/ansible/pull/75056

File: docs/docsite/sphinx_conf/2.10_conf.py (hunk 1 above)
Comment (felixfontein) [reply docs/docsite/sphinx_conf/2.10_conf.py]: Done.

File: docs/docsite/_static/core.css (hunk 2 above)
Comment (felixfontein) [reply docs/docsite/_static/core.css]: Gone after the rebase.

**Comment by felixfontein** [comment]
Marking as WIP since two more things should better happen first:
- new antsibull release with the reduced extension;
- upgrade to the latest ansible Sphinx theme.

I think it's best to continue with this once #74956 has been merged.

**Review by webknjaz — APPROVED** [review]

**Comment by webknjaz** [comment]
@felixfontein should this be unWIPed now?

**Comment by felixfontein** [comment]
@webknjaz yep it should. Done, thanks!

**Comment by felixfontein** [comment]
@abadger @webknjaz @mattclay @samccann @acozine and everyone else who was involved, thanks for reviewing, testing, ..., and merging!

**Pull request merged**